### PR TITLE
[api] Support staking action in web3

### LINKED ISSUE #2922: [api] Support other types of actions in Web3
**What would you like to be added**:

Support other types of action (like staking and rewarding) in web3
 - Transaction Object
 - Gas estimation
 - SendAction

### PR DESCRIPTION
**Pull request by Liuhaai**
fix #2922 

 - Enable staking actions able to be boardcasted to network by `sendRawTransaction()`
 - Enable staking action queryable by its act hash
 - The gas estimation of staking actions has been implemented in #3174

### CODE REVIEW

File: action/builder.go [hunk 1]
```diff
@@ -117,8 +117,8 @@ func (b *EnvelopeBuilder) SetGasPriceByBytes(buf []byte) *EnvelopeBuilder {
 	return b
 }
 
-// SetAction sets the action payload for the Envelope Builder is building.
-func (b *EnvelopeBuilder) SetAction(action actionPayload) *EnvelopeBuilder {
```
Comment (Liuhaai) [inline action/builder.go:121]: SetAction() couldn't be called outside of package

File: api/coreservice.go [hunk 2]
```diff
@@ -697,9 +705,13 @@ func (core *coreService) ReceiptByActionHash(h hash.Hash256) (*action.Receipt, e
 
 	actIndex, err := core.indexer.GetActionIndex(h[:])
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(ErrNotFound, err.Error())
+	}
+	receipt, err := core.dao.GetReceiptByActionHash(h, actIndex.BlockHeight())
+	if err != nil {
+		return nil, errors.Wrap(ErrNotFound, err.Error())
```
Comment (dustinxie) [inline api/coreservice.go:712]: this is not necessary?

File: api/coreservice.go [hunk 3]
```diff
@@ -993,16 +1005,17 @@ func (core *coreService) ActionByActionHash(h hash.Hash256) (action.SealedEnvelo
...
 	blk, err := core.dao.GetBlockByHeight(actIndex.BlockHeight())
 	if err != nil {
-		return action.SealedEnvelope{}, hash.ZeroHash256, 0, 0, err
+		return action.SealedEnvelope{}, hash.ZeroHash256, 0, 0, errors.Wrap(ErrNotFound, err.Error())
 	}
-
 	selp, index, err := core.dao.GetActionByActionHash(h, actIndex.BlockHeight())
-	return selp, blk.HashBlock(), actIndex.BlockHeight(), index, err
+	if err != nil {
+		return action.SealedEnvelope{}, hash.ZeroHash256, 0, 0, errors.Wrap(ErrNotFound, err.Error())
```
Comment (dustinxie) [inline api/coreservice.go:1016]: not necessary?

File: api/web3server.go [hunk 4]
```diff
@@ -631,42 +609,33 @@ func (svr *Web3Server) getTransactionReceipt(in interface{}) (interface{}, error
...
-
-	// acquire action receipt by action hash
 	actHash, err := hash.HexStringToHash256(util.Remove0xPrefix(actHashStr))
 	if err != nil {
 		return nil, errors.Wrapf(errUnkownType, "actHash: %s", actHashStr)
 	}
-	receipt, blkHash, err := svr.coreService.ReceiptByAction(actHash)
+
+	// acquire action receipt by action hash
+	selp, blockHash, _, _, err := svr.coreService.ActionByActionHash(actHash)
```
Comment (dustinxie) [inline api/web3server.go:618]: why need changes in `getLogs`, not the scope of this PR?

File: api/coreservice.go [hunk 5]
```diff
@@ -108,6 +108,8 @@ type (
 		ActionByActionHash(h hash.Hash256) (action.SealedEnvelope, hash.Hash256, uint64, uint32, error)
 		// ActionsByBlock returns all actions in a block
 		ActionsByBlock(blkHash string, start uint64, count uint64) ([]*iotexapi.ActionInfo, error)
+		// ActionsInBlockByHash returns all actions in a block
+		ActionsInBlockByHash(string) ([]action.SealedEnvelope, []*action.Receipt, error)
```
Comment (dustinxie) [inline api/coreservice.go:112]: not the scope of this PR?

File: api/web3server.go [hunk 6]
```diff
@@ -468,56 +469,22 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
-	}
+	elp := (&action.EnvelopeBuilder{}).SetVersion(1).SetNonce(tx.Nonce()).SetGasLimit(tx.Gas()).SetGasPrice(tx.GasPrice()).
+		SetChainID(svr.coreService.ChainID()).SetAction(act).Build()
+	selp := action.NewSealedEnvelope(
+		elp,
+		iotextypes.Encoding_ETHEREUM_RLP,
+		config.EVMNetworkID(),
+		pubkey,
+		sig,
+	)
```
Comment (dustinxie) [inline api/web3server.go:485]: here should call `NewStakingActionFromABIBinary()` to check web3 staking tx, and wrap into the execution (with special address)?

File: action/sealedenvelope.go [hunk 7]
```diff
@@ -24,6 +24,19 @@ type SealedEnvelope struct {
 	signature    []byte
 }
 
+// NewSealedEnvelope creates a new SealedEnvelope
+func NewSealedEnvelope(elp Envelope, encoding iotextypes.Encoding,
```
Comment (dustinxie) [inline action/sealedenvelope.go:28]: probably not the way we want to go about dealing `SealedEnvelope`

i think for this PR, should wrap web3 staking into the special-addressed execution and send it. See comments below. If we want to change to `sendRaw(selp SealedEnvelope)`, let's have another PR

File: api/web3server.go (hunk 4 above)
Comment (Liuhaai) [reply api/web3server.go:615]: `getTransactionFromActionInfo` is changed

File: action/sealedenvelope.go (hunk 7 above)
Comment (Liuhaai) [reply action/sealedenvelope.go]: after discusion,  ` NewSealedEnvelope` is removed

File: action/rlp_tx_test.go [hunk 8]
```diff
@@ -359,8 +359,7 @@ func convertToNativeProto(tx *types.Transaction, actType string) *iotextypes.Act
 		elpBuilder.SetAction(ex)
 	case "stakeCreate", "stakeAddDeposit", "changeCandidate", "unstake", "withdrawStake", "restake",
 		"transferStake", "candidateRegister", "candidateUpdate":
-		act, _ := NewStakingActionFromABIBinary(tx.Data())
-		elpBuilder.SetAction(act.(actionPayload))
+		elpBuilder.BuildStakingAction(tx.Data())
```
Comment (CoderZhi) [inline action/rlp_tx_test.go:362]: return ?

File: api/web3server_utils.go [hunk 9]
```diff
@@ -315,42 +300,41 @@ func (svr *Web3Server) parseBlockRange(fromStr string, toStr string) (from uint6
...
+			return elpBuilder.SetAction(exec).Build(), nil
 		}
-		return action.NewTransfer(tx.Nonce(), tx.Value(), to, tx.Data(), tx.Gas(), tx.GasPrice())
+		tsf, err := action.NewTransfer(tx.Nonce(), tx.Value(), to, tx.Data(), tx.Gas(), tx.GasPrice())
+		if err != nil {
+			return nil, err
+		}
+		return elpBuilder.SetAction(tsf).Build(), nil
 	}
 }
```
Comment (CoderZhi) [inline api/web3server_utils.go:339]: it is better to unify the interfaces of elp builder as follows:
```
  if to == address.StakingProtocolAddr {
    return elpBuilder.BuildStakingAction(tx)
  }
  isContract, err := ...
  if err != nil {
    return nil, err
  }
  if isContract {
    return elpBuilder.BuildExecution(tx)
  }

  return elpBuilder.BuildTransfer(tx)
```

File: action/builder.go [hunk 10]
```diff
@@ -139,3 +143,47 @@ func (b *EnvelopeBuilder) Build() Envelope {
 	}
 	return &b.elp
 }
+
+// BuildStakingAction loads staking action into envelope from abi-encoded data
+func (b *EnvelopeBuilder) BuildStakingAction(data []byte) (Envelope, error) {
```
Comment (CoderZhi) [inline action/builder.go:148]: types.Transaction may be easier to understand.

File: api/web3server.go (hunk 4 above)
Comment (dustinxie) [reply api/web3server.go:612]: why `getTransactionFromActionInfo` is changed then? is that needed for the web3 staking change? my understanding 
 of the necessary change is:
1. other than transfer, execution, check if the txData is a wrapped staking action
2. build the staking action as we discussed

File: api/web3server_utils.go [hunk 11]
```diff
@@ -203,98 +202,84 @@ func (svr *Web3Server) getBlockWithTransactions(blkMeta *iotextypes.BlockMeta, i
 	}, nil
 }
 
-func (svr *Web3Server) getTransactionFromActionInfo(actInfo *iotexapi.ActionInfo) (transactionObject, error) {
-	if actInfo.GetAction() == nil || actInfo.GetAction().GetCore() == nil {
-		return transactionObject{}, errNullPointer
+func (svr *Web3Server) getTransactionFromActionInfo(blkHash string, selp action.SealedEnvelope, receipt *action.Receipt) (transactionObject, error) {
```
Comment (dustinxie) [inline api/web3server_utils.go:205]: if we want to change input type to `selp`, let's do in another PR

File: api/web3server_utils.go [hunk 12]
```diff
@@ -203,98 +202,84 @@ func (svr *Web3Server) getBlockWithTransactions(blkMeta *iotextypes.BlockMeta, i
...
 		// TODO: the value is the same as Babel's. It will be corrected in next pr
 		StandardV: uint64ToHex(vVal),
 		Creates:   create,
 		ChainID:   uint64ToHex(uint64(svr.coreService.EVMNetworkID())),
-		PublicKey: byteToHex(actInfo.Action.SenderPubKey),
+		PublicKey: byteToHex(selp.SrcPubkey().Bytes()),
 	}, nil
 }
 
+func getRecipientAndContractAddrFromAction(selp action.SealedEnvelope, receipt *action.Receipt) (*string, *string, error) {
```
Comment (dustinxie) [inline api/web3server_utils.go:258]: if we want to change input type to selp, let's do in another PR

File: action/builder.go (hunk 10 above)
Comment (Liuhaai) [reply action/builder.go:148]: ok. `elpBuilder.BuildExecution(tx)`, `elpBuilder.BuildTransfer(tx)`, `elpBuilder.BuildStakingAction(tx)` looks more clear for me

File: api/web3server_utils.go [hunk 13]
```diff
@@ -203,98 +202,84 @@ func (svr *Web3Server) getBlockWithTransactions(blkMeta *iotextypes.BlockMeta, i
...
+	// recipient is empty when contract is created
+	var to, create *string
+	if exec, ok := selp.Action().(*action.Execution); ok && len(exec.Contract()) == 0 {
+		addr, err := ioAddrToEthAddr(receipt.ContractAddress)
+		if err != nil {
+			return nil, nil, err
+		}
+		to, create = nil, &addr
```
Comment (CoderZhi) [inline api/web3server_utils.go:266]: early return

**Review by CoderZhi — COMMENTED** [review]
the others look ok

File: api/web3server.go [hunk 14]
```diff
@@ -744,18 +707,15 @@ func (svr *Web3Server) getTransactionByBlockHashAndIndex(in interface{}) (interf
 	if err != nil {
 		return nil, err
 	}
-	actionInfos, err := svr.coreService.ActionsByBlock(util.Remove0xPrefix(blkHash), idx, 1)
+	blkHash = util.Remove0xPrefix(blkHash)
+	selps, receipts, err := svr.coreService.ActionsInBlockByHash(blkHash)
+	if errors.Cause(err) == ErrNotFound || idx >= uint64(len(receipts)) || len(receipts) == 0 {
```
Comment (dustinxie) [inline api/web3server.go:712]: can we use existing
```
st, ok := status.FromError(err)
		if ok && st.Code() == codes.NotFound {
			return nil, nil
		}
```
instead of introducing a new `ErrNotFound`?

File: api/web3server_utils.go [hunk 15]
```diff
@@ -315,43 +300,36 @@ func (svr *Web3Server) parseBlockRange(fromStr string, toStr string) (from uint6
...
+func (svr *Web3Server) ethTxToEnvelope(tx *types.Transaction) (action.Envelope, error) {
+	to := ""
+	if tx.To() != nil {
+		ioAddr, _ := address.FromBytes(tx.To().Bytes())
+		to = ioAddr.String()
 	}
-	ioAddr, err := address.FromString(addr)
-	if err != nil {
-		return false, err
+	elpBuilder := (&action.EnvelopeBuilder{}).SetVersion(1).SetChainID(svr.coreService.ChainID())
```
Comment (dustinxie) [inline api/web3server_utils.go:309]: remove `SetVersion(1)`,  it will be set inside `BuildTransfer()`

File: api/web3server.go (hunk 14 above)
Comment (Liuhaai) [reply api/web3server.go:712]: no. codes.NotFound will be removed in the coreservice

File: action/builder.go [hunk 16]
```diff
@@ -139,3 +146,84 @@ func (b *EnvelopeBuilder) Build() Envelope {
 	}
 	return &b.elp
 }
+
+// BuildTransfer loads transfer action into envelope
+func (b *EnvelopeBuilder) BuildTransfer(tx *types.Transaction) (Envelope, error) {
+	b.elp.nonce = tx.Nonce()
+	b.elp.gasPrice = new(big.Int).Set(tx.GasPrice())
+	b.elp.gasLimit = tx.Gas()
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply action/builder.go:154]: could you suggest a func name ?

File: action/builder.go [hunk 17]
```diff
@@ -139,3 +146,84 @@ func (b *EnvelopeBuilder) Build() Envelope {
...
+	tsf, err := NewTransfer(tx.Nonce(), tx.Value(), getRecipientAddr(tx.To()), tx.Data(), tx.Gas(), tx.GasPrice())
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply action/builder.go:155]: I think it's not a big issue here

File: action/builder.go [hunk 18]
```diff
@@ -183,6 +185,10 @@ func (b *EnvelopeBuilder) BuildExecution(tx *types.Transaction) (Envelope, error
 
 // BuildStakingAction loads staking action into envelope from abi-encoded data
 func (b *EnvelopeBuilder) BuildStakingAction(tx *types.Transaction) (Envelope, error) {
+	stakingAddr, _ := address.FromString(address.StakingProtocolAddr)
+	if !bytes.Equal(tx.To().Bytes(), stakingAddr.Bytes()) {
+		return nil, errors.New("recipient addr doesn't match")
```
Comment (dustinxie) [inline action/builder.go:190]: can return `ErrInvalidAct`

**Review by dustinxie — APPROVED** [review]

File: action/builder.go [hunk 19]
```diff
@@ -139,3 +148,91 @@ func (b *EnvelopeBuilder) Build() Envelope {
...
+	if err != nil {
+		return nil, err
+	}
+	b.elp.payload = exec
+	return b.build(), nil
+}
+
+// BuildStakingAction loads staking action into envelope from abi-encoded data
+func (b *EnvelopeBuilder) BuildStakingAction(tx *types.Transaction) (Envelope, error) {
+	stakingAddr, _ := address.FromString(address.StakingProtocolAddr)
```
Comment (CoderZhi) [inline action/builder.go:193]: two alternative ways:
1. convert in `init` function
2. convert `tx.TO()` to address.Address, and then compare to `address.StakingProtocolAddr`

**Review by CoderZhi — APPROVED** [review]

**Review by Newman1515 — COMMENTED** [review]
Got it

**Review by Newman1515 — COMMENTED** [review]
Ooo
